crypto/engine/eng_devcrypto.c

```diff
@@ -587,8 +924,121 @@ static int devcrypto_digests(ENGINE *e, const EVP_MD **digest,
     return *digest != NULL;
 }
 
+
 #endif
 
+/******************************************************************************
+ *
+ * CONTROL COMMANDS
+ *
+ *****/
+
+#define DEVCRYPTO_CMD_USE_SOFTDRIVERS ENGINE_CMD_BASE
+#define DEVCRYPTO_CMD_CIPHERS (ENGINE_CMD_BASE + 1)
+#define DEVCRYPTO_CMD_DIGESTS (ENGINE_CMD_BASE + 2)
+#define DEVCRYPTO_CMD_DUMP_INFO (ENGINE_CMD_BASE + 3)
+
+static const ENGINE_CMD_DEFN devcrypto_cmds[] = {
+   {DEVCRYPTO_CMD_USE_SOFTDRIVERS,
+    "USE_SOFTDRIVERS",
+    "specifies whether to use software (not accelerated) drivers (" STR(DEVCRYPTO_REQUIRE_ACCELERATED)
+        "=use only accelerated drivers, " STR(DEVCRYPTO_USE_SOFTWARE) "=allow all drivers, "
+        STR(DEVCRYPTO_REJECT_SOFTWARE) "=use if acceleration can't be determined [default=2])",
+    ENGINE_CMD_FLAG_NUMERIC},
+
+   {DEVCRYPTO_CMD_CIPHERS,
+    "CIPHERS",
+    "either ALL, NONE, or a comma-separated list of ciphers to enable [default=ALL]",
+    ENGINE_CMD_FLAG_STRING},
+
+#ifdef IMPLEMENT_DIGEST
+   {DEVCRYPTO_CMD_DIGESTS,
+    "DIGESTS",
+    "either ALL, NONE, or a comma-separated list of digests to enable [default=ALL]",
+    ENGINE_CMD_FLAG_STRING},
+#endif
+
+   {DEVCRYPTO_CMD_DUMP_INFO,
+    "DUMP_INFO",
+    "dump info about each algorithm to stderr; use 'openssl engine -pre DUMP_INFO devcrypto'",
+    ENGINE_CMD_FLAG_NO_INPUT},
+
+   {0, NULL, NULL, 0}
+};
+
+static int devcrypto_ctrl(ENGINE *e, int cmd, long i, void *p, void (*f) (void))
+{
+    switch(cmd) {
+    case DEVCRYPTO_CMD_USE_SOFTDRIVERS:
+        switch(i) {
+            case DEVCRYPTO_REQUIRE_ACCELERATED:
+            case DEVCRYPTO_USE_SOFTWARE:
+            case DEVCRYPTO_REJECT_SOFTWARE:
+                break;
+            default:
+                fprintf(stderr, "devcrypto: invalid value (%ld) for USE_SOFTDRIVERS\n", i);
+            return 0;
+        }
+        if (use_softdrivers == i)
+            return 1;
+        use_softdrivers = i;
+#ifdef IMPLEMENT_DIGEST
+        rebuild_known_digest_nids();
+#endif
+        rebuild_known_cipher_nids();
+        return 1;
+
+    case DEVCRYPTO_CMD_CIPHERS:
+        if (p == NULL)
+            return 1;
+        if (strcasecmp((const char *)p, "ALL") == 0) {
+            cryptodev_enable_all_ciphers();
+            goto reload_ciphers;
+        }
+        cryptodev_disable_all_ciphers();
+        if (strcasecmp((const char*)p, "NONE") == 0)
+            goto reload_ciphers;
+        else if (!CONF_parse_list(p, ',', 1, cryptodev_enable_cipher_cb, NULL)) {
+            cryptodev_enable_all_ciphers();
```

Comment:
Waaaaitaminutehere!  So all I have to do is send an invalid string, and *poof*, all ciphers are enabled, even those that were previously disabled?

Reply:
To be honest, I haven't researched what would cause this call to fail, but a bad cipher won't.  I tested it with neither a inexistant cipher, nor a valid, but unsupported one will fail it.  My thought was it was better to have them all  than none.

Reply:
With just a quick look at the callbacks, I'm going to guess that `""` (empty string), `","` (just a comma) or `,,` anywhere in the string might give you some [ahem] interesting results.

Reply:
I guess I'm failing it with an empty string, and I shouldn't.  I didn't think of it as a command really; more like a configuration file.  Set it and forget it.  Does it get much use outside of this?

Reply:
I have *no* idea about actual usages, but suffice to say that it's possible, and it's also possible that someone misspells a config line with an extra comma (which will give you a length zero call of the callback).

Reply:
I've fixed the empty string problem.  I will tackle moving the enabled parameter to a new table & saving previous state as soon as I am able.

Reply:
I have finished moving the list to a separate table.  Now I can clear it with a single memset. :grinning: